### ISSUE TITLE
Add function to set locale globally

### ISSUE THREAD

**Issue by moimael** [date-fns/date-fns#816]
Hi,

It would be great to have the possibility to call a function that set the locale globally, instead of having to wrap date-fns functions. Something similar to `moment.locale()` where we would pass the imported locale.

Thanks !

**Comment by moimael** [comment]
Is it a functionality that the core team would agree to have ?

**Comment by moimael** [comment]
We are still blocked from migrating from moment to date-fns because of this, is it planned in upcoming releases ?

**Comment by TheAifam5** [comment]
I would be happier with a global configuration for date-fns :) Not only locale but default formats, and everything which is set currently "by-default".

**Comment by nghiepit** [comment]
+1

**Comment by Brunohes** [comment]
+1

**Comment by olee** [comment]
I think it would be great to add it kinda like this:
```js
import { formatDistance } from 'date-fns'
// Require Esperanto locale
import { setDefaultLocale } from 'date-fns/locale'
import { eo } from 'date-fns/locale'
setDefaultLocale(eo);
```

This would still preserve the benefit of only importing used locales when bundling but also allows setting a default locale.
This is especially important when using date-fns in other libraries, because each one would have to do their own implementation of a default locale right now and a project using such libraries would have to set the default locale many times then.

**Comment by DaniilVysotskiy** [comment]
I may bump this issue once again.

**Comment by MartinDevillers** [comment]
+1 I know it's not very "pure" to do a global init somewhere in your code, but for the purposes of localization it's just really handy to set the user's locale once and then not having to worry about it anywhere else in your code. Without this feature, I have to wrap all locale aware methods from this library (or wire in the locale manually each time I use such a method-- ugh).

**Comment by olee** [comment]
Same reason here.
Even though you could argue like "this could break when used in libraries" etc etc, there are soo many cases where this makes so much sense like even just when you want to use it in a small nodeJs application or whatever. 
Also having to wrap it each time will just create a mess when you write libraries using date-fns, because every library will have to use it's own wrapping code which is just useless boilerplate and makes everything quite cumbersome.

Also you could argue: Why can English localization be the default, but not other locales?

**Comment by joaovbibiano** [comment]
+1

**Comment by billyjov** [comment]
+1

**Comment by xsigmar** [comment]
+1

**Comment by syrepol** [comment]
+1

**Comment by olee** [comment]
@leshakoss @kossnocorp would you mind looking into this please?
Right now this issue is a huge blocker for many users.
I know that if you are American, you this might not be an issue, but for users from other countries and international applications this is quite annoying.

**Comment by adambrgmn** [comment]
Since this issue comes up pretty high in search results on how to set default locale with `date-fns` I thought I could provide my solution:

```ts
import {
  differenceInCalendarWeeks as _differenceInCalendarWeeks,
  eachWeekOfInterval as _eachWeekOfInterval,
  endOfWeek as _endOfWeek,
  format as _format,
  formatDistance as _formatDistance,
  formatDistanceStrict as _formatDistanceStrict,
  formatDistanceToNow as _formatDistanceToNow,
  formatDistanceToNowStrict as _formatDistanceToNowStrict,
  formatDuration as _formatDuration,
  formatRelative as _formatRelative,
  getWeek as _getWeek,
  getWeekOfMonth as _getWeekOfMonth,
  getWeeksInMonth as _getWeeksInMonth,
  getWeekYear as _getWeekYear,
  isMatch as _isMatch,
  isSameWeek as _isSameWeek,
  isThisWeek as _isThisWeek,
  lastDayOfWeek as _lastDayOfWeek,
  parse as _parse,
  setDay as _setDay,
  setWeek as _setWeek,
  setWeekYear as _setWeekYear,
  startOfWeek as _startOfWeek,
  startOfWeekYear as _startOfWeekYear,
} from 'date-fns';
import { enUS } from 'date-fns/locale';

export * from 'date-fns';

let defaultLocale: { current: Locale } = { current: enUS };

export function setDefaultLocale(next: Locale) {
  defaultLocale.current = next;
}

export function resetDefaultLocale(next: Locale) {
  setDefaultLocale(enUS);
}

export const differenceInCalendarWeeks = wrap(_differenceInCalendarWeeks);
export const eachWeekOfInterval = wrap(_eachWeekOfInterval);
export const endOfWeek = wrap(_endOfWeek);
export const format = wrap(_format);
export const formatDistance = wrap(_formatDistance);
export const formatDistanceStrict = wrap(_formatDistanceStrict);
export const formatDistanceToNow = wrap(_formatDistanceToNow);
export const formatDistanceToNowStrict = wrap(_formatDistanceToNowStrict);
export const formatDuration = wrap(_formatDuration);
export const formatRelative = wrap(_formatRelative);
export const getWeek = wrap(_getWeek);
export const getWeekOfMonth = wrap(_getWeekOfMonth);
export const getWeeksInMonth = wrap(_getWeeksInMonth);
export const getWeekYear = wrap(_getWeekYear);
export const isMatch = wrap(_isMatch);
export const isSameWeek = wrap(_isSameWeek);
export const isThisWeek = wrap(_isThisWeek);
export const lastDayOfWeek = wrap(_lastDayOfWeek);
export const parse = wrap(_parse);
export const setDay = wrap(_setDay);
export const setWeek = wrap(_setWeek);
export const setWeekYear = wrap(_setWeekYear);
export const startOfWeek = wrap(_startOfWeek);
export const startOfWeekYear = wrap(_startOfWeekYear);

function wrap<T extends (...args: any[]) => any>(fn: T): T {
  let wrapped = (...args: Parameters<T>) => {
    let length = fn.length;
    let newArgs = new Array(length);
    for (let i = 0; i < length; i++) {
      if (i === length - 1) {
        newArgs[i] = { locale: defaultLocale.current, ...args[i] };
      } else {
        newArgs[i] = args[i];
      }
    }

    return fn(...newArgs);
  };

  return wrapped as T;
}
```

What happens here is that I import all methods where it's possible to provide `locale` as part of the last options argument (found from a [quick search in the `date-fns` code base](https://github.com/date-fns/date-fns/search?q=defaultLocale)). Then I wrap these methods in a higher order component that will append the current locale as the last argument, without overriding any previously passed locale.

As seen after the imports I also export two methods to set and reset the default locale.

This solution preserves types if you're a fan of typescript. And it works thanks to the strict standards setup by `date-fns` where locale is _always_ passed as part of a config object as the last argument to the methods.

**Comment by olee** [comment]
The issue is not about wrapping date-fns - pretty much every developer knows how to do that.
The issue is that wrapping it breaks consistency in the application if there are multiple libraries depending on date-fns and it requires a lot of duplicate code.
The only alternative would be releasing a common wrapper as a separete npm package and trying to make library depending on date-fns to use this wrapper instead ... which just sounds horrible.

**Comment by fturmel** [comment]
This would be a very useful feature. I'm willing to make it happen, but would like an official go-ahead from the maintainers before I work on a PR given that this is a very old issue and #1885 fizzled out. @tan75 @kossnocorp

**Comment by mellamopepino** [comment]
Hi, @tan75 @kossnocorp can you please review this issue? It blocks a lot of users.
Not everyone speaks english, so this is quite annoying.

**Comment by pascalvos** [comment]
i made a working solution using proxy settings but not having it inside the library kinda sucks 
check out the demo here https://codesandbox.io/s/lingering-thunder-jr73k

**Comment by Fanna1119** [comment]
i just found myself needing this too.
sometimes a users browser/system locale is different than what you would like for your app/site. would be handy when working with different timezones and all.
Being able to set this globally would take a lot of headache away.

**Comment by ShivamJoker** [comment]
Hello any updates on this ?

**Comment by DesKevinMendez** [comment]
Set a local as globally is not the expected behavior of library I prefer write in some file only what I need, for example, I need convert a Date to specific format; write a function for that. 

Do I really need the locale globally? The answer is not, I just need set local to format.

My recomendation is only set locale in what you need. No more. 

Here I wrote some example.

```
import { parseISO, format } from 'date-fns'
import { es } from 'date-fns/locale'

export const stringDateToJsDate = (date: string): Date => {
  return parseISO(date)
}
export const svDate = (dateConvert: Date): string =>
  format(dateConvert, 'EEE dd yyyy HH:mm a', {
    locale: es
  })
```

I understand; exists cases what we need added a globally configuration, but think, do you really need added it?

**Comment by syrepol** [comment]
To those who struggle with this issue, probably this can help. We use a similar approach to what @DesKevinMendez mentioned

<img width="763" alt="Screenshot 2021-11-04 at 12 28 07" src="https://user-images.githubusercontent.com/22783068/140298310-87a75894-f78d-495d-b996-c41e23a6b47b.png">
<img width="760" alt="Screenshot 2021-11-04 at 12 28 45" src="https://user-images.githubusercontent.com/22783068/140298539-c4f0a6b7-c700-4f62-b0e5-696b0407a1b2.png">

So instead of cluttering otherwise functional library with some global state, there is a wrapper that we use anywhere formatting is needed instead of date-fns's format().

**Comment by mellamopepino** [comment]
I dont see how what @DesKevinMendez  and @syrepol suggested is helpful for all the users that commented in here.
The first option: set the locale where you need it. Well... I need it everywhere, how is manually setting it every single time a good solution?
The second option: wrap it where you need it. Again, I need it everywhere, is the same problem that the first option.

The idea in the users asking this feature is that you can configure this setting one single time and then you are all set to format with the locale you need.
Because the alternative is setting it manually every time, and this is more prone to error.
Am I missing something in these suggestions?

**Comment by mellamopepino** [comment]
> @mellamopepino
> 
> > Am I missing something in these suggestions?
> 
> Yes, I think you are 😁
> 
> > The idea in the users asking this feature is that you can configure this setting one single time and then you are all set to format with the locale you need.
> 
> Exactly how it works in the example I provided above. You do this setup once and then use this wrapper function everywhere. I think it's even beneficial since you have an abstraction over a particular dates library which makes some theoretical migrations between them much easier.
> 
> I doubt global locale settings will ever be possible in this library since it's not pure, besides, this change isn't even required to solve the issue, as we already clarified

Yeah, I didn't miss that. You are still having to use the wrap function everywhere. Or wrap all the library functions with it and then export every single wraped fuction. Is the same, those options are still a hassle.

And it makes theorical migrations easier only compared to the no-wrap option, but not compared to the global setting one, since you will still have to change the settings or hope that the settings in another library are exactly the same.

Last, I didn't saw this library code to be honest, so I don't know if its posible or not. But I still believe is the best option, if they can't make it I'm happy with a "sry, but is not possible".
Why you say this change isn't required to solve the issue? I still didn't see a good solution without that change.

**Comment by syrepol** [comment]
> Last, I didn't saw this library code to be honest, so I don't know if its posible or not. But I still believe is the best option, if they can't make it I'm happy with a "sry, but is not possible".

You don't have to wrap **every** function, just `parse` and `format`, probably a couple more on top of it. In our project  wrapping just those two solved all of the issues with locale and we have fairly complex dates-related calculations

**Comment by olee** [comment]
@syrepol there are a few issues in your reasoning which you missed:
- There **is** already a global state of locale - it is just that it is hardcoded to be the **english locale**. So if we argue in your way to not have global state, then the locale would have to become a required parameter and I would like to see the backlash if this would be done...
- While wrapping the required functions is indeed a working solution, it is the worst solution imaginable:
  1. It requires writing a lot of useless code just to replace something which should be a changeable variable instead of a constant ("english") within date-fns
  2. When using typescript, also getting all the typings correct for the wrappers becomes even more effort
  3. When other libraries want to use date-fns they won't use the wrappers and have to rely on separately passed in options again or even define their own wrappers => duplicate code

About this not being "possible":  
It is actually super simple by just replacing all places where `import defaultLocale from '../locale/en-US/index'` is used [such as here](https://github.com/date-fns/date-fns/blob/master/src/format/index.js#L354) to some `../globalLocale` file which would just contain something like this:
```js
import defaultLocale from '../locale/en-US/index'
export let globalLocale = defaultLocale;
export function setGlobalLocale(locale) {
  globalLocale = locale || defaultLocale;
}
```

If you consider all of this simple required code change which would not affect date-fns in any negative way, there's no good reason against it.

So once more in good hope:  
Please @kossnocorp, could you give a "go ahead" to this issue so a PR could be created? Not having this feature is really infuriating if you're targeting non-english audiences and it would simplify using this library a lot.

**Comment by happylolonly** [comment]
Yes, guys, please add this functionality 🙏

**Comment by fturmel** [comment]
Took a shot at implementing this. Please have a look at the PR if you can! #2816

**Comment by juliomac** [comment]
+1

**Comment by leshakoss** [comment]
Released in v2.29: https://date-fns.org/v2.29.0/docs/setDefaultOptions

**Comment by marchiartur** [comment]
> Released in v2.29: https://date-fns.org/v2.29.0/docs/setDefaultOptions

why do I need to import the locale? why can't I just set the locale passing the code string?
